Add sanity checks for CodeBlock::destroy().

MenloDorian opened the pull request:
#### 8d86a7bac10773a90ff14e768cfb3d5ecd9967d2
<pre>
Add sanity checks for CodeBlock::destroy().
<a href="https://bugs.webkit.org/show_bug.cgi?id=291532">https://bugs.webkit.org/show_bug.cgi?id=291532</a>
<a href="https://rdar.apple.com/143559929">rdar://143559929</a>

Reviewed by Keith Miller and Yijia Huang.

This is temporarily needed for crash analysis on macOS.

* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::CodeBlock):
(JSC::CodeBlock::setBaselineJITData):
(JSC::CodeBlock::~CodeBlock):
* Source/JavaScriptCore/bytecode/CodeBlock.h:
(JSC::CodeBlock::CrashChecker::hash):
(JSC::CodeBlock::CrashChecker::get):
(JSC::CodeBlock::CrashChecker::set):
(JSC::CodeBlock::CrashChecker::value const):
(JSC::CodeBlock::setDFGJITData):
(JSC::CodeBlock::setBaselineJITData): Deleted.

Canonical link: <a href="https://commits.webkit.org/293712@main">https://commits.webkit.org/293712@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e0645cacf8ce04a6f9fe04bd7629a4b291dffcd3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99664 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19313 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9575 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104793 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50258 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19602 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27747 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75868 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32962 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102671 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14928 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89993 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56228 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14728 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7979 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49620 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/92342 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/84687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8064 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107153 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/98281 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26778 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19550 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84826 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27142 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86197 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84344 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21420 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29017 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6725 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20569 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26719 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31922 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/121898 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26534 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34048 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29849 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28103 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/bytecode/CodeBlock.cpp:
```diff
@@ -824,9 +839,29 @@ void CodeBlock::setupWithUnlinkedBaselineCode(Ref<BaselineJITCode> jitCode)
 CodeBlock::~CodeBlock()
 {
+    auto& cc = m_checker;
+    if (cc.isEnabled) {
```
use `UNLIKELY(cc.isEnabled)`?

MenloDorian replied:
`isEnabled` is a `static constexpr bool`.  This code will either be elided or it will have to always be executed.  Hence, there's no point in making it `UNLIKELY`.  I only expressed it as `cc.isEnabled` for brevity.  Alternately, I could have used `CrashChecker:: isEnabled` and use `if constexpr` here instead.  Regardless, the compiler should just optimize this out trivially.

hyjorc1 commented on Source/JavaScriptCore/bytecode/CodeBlock.cpp:
```diff
@@ -824,9 +839,29 @@ void CodeBlock::setupWithUnlinkedBaselineCode(Ref<BaselineJITCode> jitCode)
     VM& vm = *m_vm;
 
     if (JITCode::isBaselineCode(jitType())) {
+        if (cc.isEnabled && m_jitData) {
```
Ditto.

hyjorc1 commented on Source/JavaScriptCore/bytecode/CodeBlock.cpp:
```diff
@@ -886,6 +921,20 @@ CodeBlock::~CodeBlock()
     // destructors.
 
 #if ENABLE(JIT)
+    if (cc.isEnabled && m_jitData) {
```
Ditto.

hyjorc1 commented on Source/JavaScriptCore/bytecode/CodeBlock.cpp:
```diff
@@ -905,6 +954,53 @@ CodeBlock::~CodeBlock()
         }
     }
 #endif // ENABLE(JIT)
+    if (cc.isEnabled) {
```
Ditto.

kmiller68 reviewed (approved):
r=me

hyjorc1 reviewed (approved):
r=me

Commit:
Add sanity checks for CodeBlock::destroy().
https://bugs.webkit.org/show_bug.cgi?id=291532
rdar://143559929

Reviewed by Keith Miller and Yijia Huang.

This is temporarily needed for crash analysis on macOS.

* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::CodeBlock):
(JSC::CodeBlock::setBaselineJITData):
(JSC::CodeBlock::~CodeBlock):
* Source/JavaScriptCore/bytecode/CodeBlock.h:
(JSC::CodeBlock::CrashChecker::hash):
(JSC::CodeBlock::CrashChecker::get):
(JSC::CodeBlock::CrashChecker::set):
(JSC::CodeBlock::CrashChecker::value const):
(JSC::CodeBlock::setDFGJITData):
(JSC::CodeBlock::setBaselineJITData): Deleted.

Canonical link: https://commits.webkit.org/293712@main
